# Read TSV file and write Arrow

Attach the packages to be used

In [1]:
using Arrow, CSV, DataFrames

Read the tab-separated-value file

In [2]:
df = CSV.read(
    "ncvoter_Statewide.txt",
    DataFrame;              # return as a DataFrame
    delim='\t',             # tab-separated values
    dateformat="mm/dd/yyyy",
    missingstring=["", " ", "##/##/####"],
    downcast=true,          # use small integer sizes if possible
    pool=0.3,               # use PooledArrays if less than 30% distinct
    stripwhitespace=true,
    stringtype=String,      # avoid complexities from special string types
)

8778585×67 DataFrame
     Row │ county_id  county_desc  voter_reg_num  ncid      last_name          ⋯
         │ Int8       String       Int32          String    String             ⋯
─────────┼──────────────────────────────────────────────────────────────────────
       1 │         1  ALAMANCE           9005990  AA56273   AABEL              ⋯
       2 │         1  ALAMANCE           9178574  AA201627  AARDEN
       3 │         1  ALAMANCE           9205561  AA216996  AARMSTRONG
       4 │         1  ALAMANCE           9048723  AA98377   AARON
       5 │         1  ALAMANCE           9019674  AA69747   AARON              ⋯
       6 │         1  ALAMANCE           9129589  AA170513  AARON
       7 │         1  ALAMANCE           9194595  BM49306   AARON
       8 │         1  ALAMANCE           9041748  AA91549   AARON
    ⋮    │     ⋮           ⋮             ⋮           ⋮              ⋮          ⋱
 8778579 │       100  YANCEY               15235  ES17216   ZITNEY             ⋯
 8778580 │       100  YANCEY               33554  ES26136   ZITNEY
 8778581 │       100  YANCEY               15223  ES17204   ZITNEY
 8778582 │       100  YANCEY               26014  ES26341   ZULICK
 8778583 │       100  YANCEY               26085  ES26380   ZULLINGER          ⋯
 8778584 │       100  YANCEY               34694  AL300586  ZURAWSKI
 8778585 │       100  YANCEY               30180  ES28936   ZURAWSKI
                                             62 columns and 8778570 rows omitted

Summary of values in table

In [3]:
desc = describe(df, :min, :max, :nunique, :nmissing, :eltype)
desc[1:25, :]

25×6 DataFrame
 Row │ variable           min               max                           nuni ⋯
     │ Symbol             Any               Any                           Unio ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ county_id          1                 100                                ⋯
   2 │ county_desc        ALAMANCE          YANCEY                        100
   3 │ voter_reg_num      0                 1000931695                        
   4 │ ncid               AA100000          ES9999                        8778
   5 │ last_name           CARROLL          ZYWOT                         5114 ⋯
   6 │ first_name          GARLANDA         ZZREAM                        3362
   7 │ middle_name         BULLABOY         ZYZO                          4615
   8 │ name_suffix_lbl    E                 X                             17
  ⋮  │         ⋮                 ⋮                       ⋮                   ⋮ ⋱
  19 │ mail_addr3         # 1708            \xc5S / 1433                  9549 ⋯
  20 │ mail_addr4         (ROOM 510)        \xd6STERREICH                 885
  21 │ mail_city           ASHEVILLE,       |                             7973
  22 │ mail_state          V                WY                            94
  23 │ mail_zipcode       0                 997095241                          ⋯
  24 │ full_phone_number     3272           9999999999                    3370
  25 │ confidential_ind   N                 Y                             2
                                                   3 columns and 10 rows omitted

In [4]:
desc[26:50,:]

25×6 DataFrame
 Row │ variable              min                      max                nuniq ⋯
     │ Symbol                Any                      Any                Union ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ registr_dt            1938-01-01               2025-01-24         23255 ⋯
   2 │ race_code             A                        W                  8
   3 │ ethnic_code           HL                       UN                 3
   4 │ party_cd              CST                      WTP                9
   5 │ gender_code           F                        U                  3     ⋯
   6 │ birth_year            1753                     2008                    
   7 │ age_at_year_end       17                       272                     
   8 │ birth_state           AK                       WY                 57
  ⋮  │          ⋮                       ⋮                     ⋮             ⋮  ⋱
  19 │ nc_senate_abbrv       1                        50                       ⋯
  20 │ nc_house_abbrv        1                        120                     
  21 │ county_commiss_abbrv  #01                      WEST               122
  22 │ county_commiss_desc   #01                      WEST COMMISSIONER  142
  23 │ township_abbrv        01                       WILL               83    ⋯
  24 │ township_desc         ANDERSON CREEK TOWNSHIP  YADKIN             123
  25 │ school_dist_abbrv     #01                      Y                  144
                                                   3 columns and 10 rows omitted

In [5]:
desc[51:end, :]

17×6 DataFrame
 Row │ variable           min                       max                        ⋯
     │ Symbol             Any                       Any                        ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │ school_dist_desc   #01                       WHITEVILLE SCHOOL DISTRICT ⋯
   2 │ fire_dist_abbrv    01                        ZION
   3 │ fire_dist_desc     ADR FIRE DISTRICT         ZIONVILLE
   4 │ water_dist_abbrv   01                        WD00
   5 │ water_dist_desc    CENTRAL                   WOODFIN WATER DISTRICT     ⋯
   6 │ sewer_dist_abbrv   0001                      LP
   7 │ sewer_dist_desc    DISTRICT 1                TUNIS
   8 │ sanit_dist_abbrv   1                         WSD
  ⋮  │         ⋮                     ⋮                           ⋮             ⋱
  11 │ rescue_dist_desc   ADR RESCUE DISTRICT       WHITEVILLE RESCUE DISTRICT ⋯
  12 │ munic_dist_abbrv   01                        ZEB
  13 │ munic_dist_desc    ABERDEEN                  ZEBULON
  14 │ dist_1_abbrv       1                         43
  15 │ dist_1_desc        PROSECUTORIAL DISTRICT 1  PROSECUTORIAL DISTRICT 9   ⋯
  16 │ vtd_abbrv          0001                      YAN4
  17 │ vtd_desc           0001                      YAN4
                                                    4 columns and 2 rows omitted

There are peculiarities in this table (birth_year of 1753, coding glitches) but we will ignore them for the time being.

In [6]:
afn = Arrow.write("ncvoter_Statewide.arrow", df);   # uncompressed Arrow file

Next, read it back in to Julia.  Because we did not use compression of the columns the file can be memory-mapped to be read, which is very fast.

In [7]:
tbl = Arrow.Table(afn)

Arrow.Table with 8778585 rows, 67 columns, and schema:
 :county_id          Int8
 :county_desc        String
 :voter_reg_num      Int32
 :ncid               String
 :last_name          String
 :first_name         Union{Missing, String}
 :middle_name        Union{Missing, String}
 :name_suffix_lbl    Union{Missing, String}
 :status_cd          String
 :voter_status_desc  String
 ⋮                   
 :sanit_dist_desc    Union{Missing, String}
 :rescue_dist_abbrv  Union{Missing, String}
 :rescue_dist_desc   Union{Missing, String}
 :munic_dist_abbrv   Union{Missing, String}
 :munic_dist_desc    Union{Missing, String}
 :dist_1_abbrv       Union{Missing, Int8}
 :dist_1_desc        Union{Missing, String}
 :vtd_abbrv          Union{Missing, String}
 :vtd_desc           Union{Missing, String}